### PR TITLE
feat(storage): Added Bulk Insertion Support (for Vertices and Edges) in AdjacencyList.hpp and GraphStatistics functionality to the codebase.

### LINKED ISSUE #45: feat(storage): Add method for bulk insertions in AdjacencyList.hpp and Graph Statistics in the codebase
The current `AdjacencyList.hpp` implementation handles operations one at a time. The `addVertex()` and `addEdge()` methods perform individual insertions without bulk/batch operation support. Also, there is no method to get storage metrics like memory usage, vertex/edge counts, or adjacency list load factors, which are essential for performance monitoring and optimization decisions.

Solution: Add `impl_addVertices() and `impl_addEdges()` methods that process multiple operations efficiently. Implement `getGraphStatistics()` method returning graphStatistics. 

@SharonIV0x86 Please assign this issue to me so that I can work on it.

### PR DESCRIPTION
**Pull request by me-hem**
This PR introduces **bulk insertion support** for vertices and edges in the adjacency list, addressing (issue #45) The new functions streamline graph construction by reducing repetitive single-insert calls.

### Changes Made

* **`AdjacencyList.hpp`**
  - Added `impl_addEdgesBatch` for inserting multiple edges in one call.
  - Added `impl_addVerticesBatch` for inserting multiple vertices in one call.

* **`AdjacencyShard.cpp`**
  - Added and refined unit tests to validate bulk insertion behavior.
  - Extended coverage for edge cases such as duplicate entries and empty input.

### Benefits

* Improves efficiency when adding large sets of edges or vertices.
* Reduces boilerplate in client code.
* Ensures consistent handling of duplicates and invalid inputs.

### Test Results
bulk_insertion

### CODE REVIEW

**Comment by SharonIV0x86** [comment]
You batch functions boils down to calling the 

```js
_adj_list[src].emplace_back(dest, weight);
```
and
```js
_adj_list[vertex] = std::vector<std::pair<VertexType, EdgeType>>();
```
in a loop, which comes out to be the same time complexity and hash table look ups as if done sequentially.
This seems good as a convenience wrapper but not as optimized as making a buffer/cache like structure and committing it directly by reserving the space.

**Comment by me-hem** [comment]
You are right - the current batch functions are convenience wrappers that do not provide algorithmic optimizations. They are designed as practical utilities for bulk data loading scenarios (CSV files, databases, etc.) where users need cleaner APIs and unified error handling rather than repetitive individual calls. While the core hash operations remain the same, the batch approach does reduce function call overhead compared to thousands of individual addVertex()/addEdge() calls. The implementation prioritizes simplicity and usability for common graph construction patterns. If you'd like, I can enhance these with actual performance optimizations (buffering, pre-allocation, reduced hash lookups) in a future PR.

**Comment by SharonIV0x86** [comment]
> You are right ..... PR.

Seems fair, we can offer ``addEdges`` and ``addVertices`` methods from the API.

```js
graph.addEdges({1, 2}, {3, 4}, {5,6});
// or
graph.addEdges({1, 2, 1}, {3, 4, 2}, {5, 6, 7});
```
and for vertices

```js
graph.addVertices(12, 3, 4, 2, 1, 6, 7);
// or
graph.addVertices({12, 3, 4,2, 1, 6, 7});
```

This will require add some Variadic complexity but ig it will be fine.

It is now up to you if you want to work with actual batch and commits with a shared buffer or keep it this way.
Also you can try to follow some optimization techniques in c++
https://cplusplus.com/reference/unordered_map/unordered_map/reserve/

**Comment by me-hem** [comment]
@SharonIV0x86 This time I want to go with `addEdges` and `addVertices` methods from the API. Will do further optimizations later.

**Comment by me-hem** [comment]
@SharonIV0x86 I was thinking about the variadic design you suggested, but it has a practical scalability issue:

```cpp
// Variadic: fine for small cases
graph.addVertices(1, 2, 3);
graph.addEdges({1, 2}, {3, 4}, {5, 6});

// But impractical for real datasets  
graph.addVertices(1, 2, 3, /* ...manually type 1000+ individual args? */);
graph.addEdges({1, 2}, {3, 4}, /* ...manually type 1000+ individual pairs? */);
```

My implementation with container approach scales naturally and allows programmatic construction:

```cpp
// For vertices
graph.addVertices({12, 3, 4, 2, 1, 6, 7});

// For edges (unweighted and weighted)
graph.addEdges({{1, 2}, {3, 4}, {5, 6}});
graph.addEdges({{1, 2, 1}, {3, 4, 2}, {5, 6, 7}});

// Can be built programmatically
std::vector<int> vertices;
for (int i = 0; i < 1000; ++i) vertices.push_back(i);
graph.addVertices(vertices);

std::vector<std::pair<int, int>> edges;
for (int i = 0; i < 999; ++i) edges.emplace_back(i, i + 1);
graph.addEdges(edges);
```

The syntax difference is minimal, but container approach enables both manual and automatic data population (including loading from files). Thoughts?

**Comment by SharonIV0x86** [comment]
@me-hem There can be multiple approaches to this. Variadic was one of them which popped into my mind at the moment and that's why I mentioned it, I am not sure of using it, I just mentioned it because I have used it before for the same thing. 


Anyways it will be done on the API level i suggest you keep this PR focused towards optimising the additions to the unordered_map in bulk and also updating the ``GraphInternalMetadata`` structure to enable statistics information.

**Comment by me-hem** [comment]
@SharonIV0x86 I have updated the PR and implemented `GraphInternalMetadata` functionalities. And I will optimize the operation in my future PRs. So, from my side this PR is done.

Note: You have linked wrong issue with this PR so please correct that (related issue is #45).

### Test results:
Stat1

Stat2

File: src/PeakStore.hpp [hunk 1]
```diff
@@ -105,6 +110,27 @@ template <typename VertexType, typename EdgeType> class PeakStore {
   getContext() const {
     return ctx;
   }
+  
+  // Method to get a summary string of statistics
+  std::string getGraphStatistics() {
```
Comment (SharonIV0x86) [inline src/PeakStore.hpp:115]: This function seems fine to just print the statistics, but we would need a more idiomatic way to let user get information about their graph, so for this PR this is fine.

**Review by SharonIV0x86 — APPROVED** [review]
Approved.

**Comment by SharonIV0x86** [comment]
@me-hem Thank you for contributing your PR has been merged.
